swupd/manifest.go

```diff
@@ -363,59 +363,100 @@ func (m *Manifest) prependFile(f *File) {
 // At this point Manifest m should only have the files that were present
 // in the chroot for that manifest. Link delta peers with the oldManifest
 // if the file in the oldManifest is not deleted or ghosted
-func (m *Manifest) linkPeersAndChange(oldManifest *Manifest) int {
+func (m *Manifest) linkPeersAndChange(oldManifest *Manifest) (int, int, int) {
+	m.sortFilesName()
+	oldManifest.sortFilesName()
+
 	changed := 0
-	for _, of := range oldManifest.Files {
-		if match := of.findFileNameInSlice(m.Files); match != nil {
+	added := 0
+	deleted := 0
+	i := 0
+	j := 0
+	for i < len(m.Files) && j < len(oldManifest.Files) {
+		nf := m.Files[i]
+		of := oldManifest.Files[j]
+		if nf.Name == of.Name {
+			// if it is the same name check if anything about the file has changed.
+			// if something has changed update the version to current version and
+			// record that the file has changed
 			if of.Status == statusDeleted || of.Status == statusGhosted {
+				// don't worry about ghosted or deleted files
+				i++
+				j++
 				continue
 			}
 
-			match.DeltaPeer = of
-			of.DeltaPeer = match
-			if !sameFile(match, of) {
-				match.Version = m.Header.Version
+			// set up peers
+			nf.DeltaPeer = of
+			of.DeltaPeer = nf
+			if !sameFile(nf, of) {
+				// if the file isn't exactly the same, record the change
+				// and update the version
+				nf.Version = m.Header.Version
 				changed++
 			} else {
-				match.Version = of.Version
+				// otherwise the version is the same as the old file
+				nf.Version = of.Version
+			}
+			// advance indices for both file lists since we had a match
+			i++
+			j++
+		} else if nf.Name < of.Name {
+			// if the file does not exist in the old manifest it is a new
+			// file in this manifest. Update the version and record that
+			// it is added.
+			nf.Version = m.Header.Version
+			added++
+			// look at next file in current manifest
+			i++
+		} else {
+			// if the file exists in the old manifest and does not *yet* exist
+			// in the new manifest, it was deleted. Set up the deleted file
+			// and prepend it to the files list via newDeleted()
+			if of.Status == statusDeleted || of.Status == statusGhosted {
+				j++
+				continue
 			}
+			m.newDeleted(of)
+			deleted++
+			// look at next file in old manifest
+			j++
+			// increment the index into m.Files as well since we increased
+			// length by prepending of
```

Comment:
This i++ looks like we are doing the wrong thing!

I think that by the time we get to line 445 we should have a a local slice of each of the following

files which exist in both the old and new manifests
files which didn't exist in the old but do exist in the new
files which did exist in the old and don't exist in the new. (this is essentially the newDeleted stuff, but I don't think we want to alter the attributes yet)
The rename detection code then works on matching things in the second and third lists, if it finds a match then it removes the pair of files from these two lists and adds it to the first.

At this point we can compute "changed", added, and deleted.

We then merge the 3 local slices to form the Files slice of the new manifest.

Reply:
I see your point for creating the three different slices so rename detection is faster... but I don't see the point of then merging these three slices at the end when we can just add them *in* the loop like we are already doing. If the three slices contain pointers to the files in the manifest.Files list then we can just update the rename status and the hash in the lists and the master list will have the changes.